api/s2n.h

```diff
@@ -3679,6 +3679,82 @@ S2N_API int s2n_offered_early_data_accept(struct s2n_offered_early_data *early_d
+S2N_API int s2n_config_set_serialized_connection_version(struct s2n_config *config, s2n_serialization_version version);
+
+/**
+ * Retrieves the length of the serialized connection retrieved from s2n_connection_serialize. Should be
+ * used to allocate enough memory for the serialized connection buffer.
+ * 
+ * @note The size of the serialized connection changes based on parameters negotiated in the TLS
+ * handshake. Do not expect the size to always remain the same.
+ * 
+ * @param conn A pointer to the connection object.
+ * @param length Output parameter where the length will be written.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on error.
+ */
+S2N_API int s2n_connection_serialization_length(struct s2n_connection *conn, uint32_t *length);
+
+/**
+ * Serializes the s2n_connection into the provided buffer.
+ * 
+ * This API takes an established s2n-tls connection object and "serializes" it
+ * into a transferable object to be sent off-box or stored. This transferable object can then
+ * be "deserialized" using the `s2n_connection_deserialize` method to instantiate an s2n-tls 
+ * connection object that can talk to the original peer with the same encryption keys.
+ *
+ * Note that this feature is dangerous because it provides cryptographic material from a TLS session
+ * in plaintext. Users MUST both encrypt and MAC the contents of the outputted material to provide
+ * secrecy and integrity if this material is transported off-box. DO NOT send this material off-box
+ * without a plan to do so.
+ * 
+ * @note You MUST call `s2n_config_set_serialized_connection_version()` on the s2n_config object used
+ * to configure this connection BEFORE calling this API.
+ * @note Call `s2n_connection_serialization_length` to retrieve the amount of memory needed for the
+ * buffer parameter.
+ * @note This API will error if the handshake is not yet complete.
+ * 
+ * @param conn A pointer to the connection object.
+ * @param buffer A pointer to the buffer where the serialized connection will be written.
+ * @param buffer_length Maximum amount of data that can be written to the buffer param.
+ */
+S2N_API int s2n_connection_serialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length);
+
+/**
+ * Deserializes the provided buffer into the `s2n_connection` parameter.
+ * 
+ * @warning s2n-tls DOES NOT check the integrity of the provided buffer. s2n-tls may successfully 
+ * deserialize a corrupted buffer which WILL cause a connection failure when attempting to resume
+ * sending/receiving encrypted data. To avoid this, it is recommended to MAC and encrypt the serialized 
+ * connection before sending it off-box and deserializing it.
+ * 
+ * @warning Only the minimal amount of information about the original TLS connection is serialized.
+ * Therefore, after deserialization, the connection will behave like a new `s2n_connection` from the 
+ * `s2n_new_connection()` call, except that it can read/write encrypted data from a peer. None of the
```

Comment:
Now you're even applying your love of "verb_noun" naming to existing APIs 😔
```suggestion
 * Therefore, after deserialization, the connection will behave like a new `s2n_connection` from the 
 * `s2n_connection_new()` call, except that it can read/write encrypted data from a peer. None of the
```